Create a simple Roberta Training model

In [ ]:
!pip install transformers -U
!pip install torch -U

In [ ]:
!apt install git-lfs

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
git-lfs is already the newest version (3.0.2-1ubuntu0.2).
0 upgraded, 0 newly installed, 0 to remove and 9 not upgraded.


In [ ]:
!pip install sentencepiece

In [ ]:
import transformers

print(transformers.__version__)

4.35.2


In [ ]:
#from transformers import *

In [ ]:
from huggingface_hub import notebook_login
#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

notebook_login()

In [ ]:
import torch
torch.cuda.is_available()

True

In [ ]:
from transformers import RobertaTokenizerFast
#Model with pre trained tokenizer: rebe121314/BPE_Tokenizer_seq2seq_smiles

In [ ]:
tokenizer = RobertaTokenizerFast.from_pretrained('rebe121314/BPE_Tokenizer_seq2seq_smiles')

loading file vocab.json from cache at None
loading file merges.txt from cache at None
loading file tokenizer.json from cache at /root/.cache/huggingface/hub/models--rebe121314--BPE_Tokenizer_seq2seq_smiles/snapshots/d41df5e6c02e380e813934b9449fc30c8fc7bcb6/tokenizer.json
loading file added_tokens.json from cache at None
loading file special_tokens_map.json from cache at /root/.cache/huggingface/hub/models--rebe121314--BPE_Tokenizer_seq2seq_smiles/snapshots/d41df5e6c02e380e813934b9449fc30c8fc7bcb6/special_tokens_map.json
loading file tokenizer_config.json from cache at /root/.cache/huggingface/hub/models--rebe121314--BPE_Tokenizer_seq2seq_smiles/snapshots/d41df5e6c02e380e813934b9449fc30c8fc7bcb6/tokenizer_config.json
The tokenizer class you load from this checkpoint is not the same type as the class this function is called from. It may result in unexpected tokenization. 
The tokenizer class you load from this checkpoint is 'PreTrainedTokenizerFast'. 
The class this function is called fr

In [ ]:
%%capture
!sudo apt -qq install git
!git config --global credential.helper store

In [ ]:
from transformers import RobertaConfig

config = RobertaConfig(
    vocab_size=8000,
    max_position_embeddings=514,
    num_attention_heads=12,
    num_hidden_layers=6,
    type_vocab_size=1,
)

In [ ]:
from transformers import RobertaForMaskedLM

model = RobertaForMaskedLM(config=config)

In [ ]:
model.num_parameters()


49668416

In [ ]:
%%time
from transformers import LineByLineTextDataset

dataset = LineByLineTextDataset(
    tokenizer=tokenizer,
    file_path="/content/smilecsv_hiv_general.txt",
    block_size=128,
)

/usr/local/lib/python3.10/dist-packages/transformers/data/datasets/language_modeling.py:119: FutureWarning: This dataset will be removed from the library soon, preprocessing should be handled with the 🤗 Datasets library. You can have a look at this example script for pointers: https://github.com/huggingface/transformers/blob/main/examples/pytorch/language-modeling/run_mlm.py
  warnings.warn(
Creating features from dataset file at /content/smilecsv_hiv_general.txt


CPU times: user 3.03 s, sys: 188 ms, total: 3.22 s
Wall time: 3.66 s


In [ ]:
from transformers import DataCollatorForLanguageModeling

data_collator = DataCollatorForLanguageModeling(
    tokenizer=tokenizer, mlm=True, mlm_probability=0.15
)

In [ ]:
!pip install accelerate

In [ ]:
from huggingface_hub import HfFolder, Repository, whoami
from accelerate import Accelerator


In [ ]:
!pip install accelerate -U

In [ ]:
!pip install transformers[torch]

In [ ]:
!nvidia-smi

Wed Nov 22 15:55:27 2023       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.105.17   Driver Version: 525.105.17   CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            Off  | 00000000:00:04.0 Off |                    0 |
| N/A   63C    P0    31W /  70W |    509MiB / 15360MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

In [ ]:

from transformers import Trainer, TrainingArguments

training_args = TrainingArguments(
    output_dir="SMILES_RoBERTa",
    overwrite_output_dir=True,
    num_train_epochs=2,
    per_device_train_batch_size=64,
    save_steps=10_000,
    save_total_limit=2,
    prediction_loss_only=True,
)

trainer = Trainer(
    model=model,
    args=training_args,
    data_collator=data_collator,
    train_dataset=dataset,
)

#The default value for the training argument `--report_to` will change in v5 (from all installed integrations to none). In v5, you will need to use `--report_to all` to get the same behavior as now. You should start updating your code and make this info disappear :-).


PyTorch: setting up devices
The default value for the training argument `--report_to` will change in v5 (from all installed integrations to none). In v5, you will need to use `--report_to all` to get the same behavior as now. You should start updating your code and make this info disappear :-).


In [ ]:
%%time
trainer.train()

***** Running training *****
  Num examples = 32,902
  Num Epochs = 2
  Instantaneous batch size per device = 64
  Total train batch size (w. parallel, distributed & accumulation) = 64
  Gradient Accumulation steps = 1
  Total optimization steps = 1,030
  Number of trainable parameters = 49,668,416




Training completed. Do not forget to share your model on huggingface.co/models =)




CPU times: user 10min 40s, sys: 1.47 s, total: 10min 42s
Wall time: 10min 46s


TrainOutput(global_step=1030, training_loss=1.2289109313372268, metrics={'train_runtime': 645.6688, 'train_samples_per_second': 101.916, 'train_steps_per_second': 1.595, 'total_flos': 1807646409192192.0, 'train_loss': 1.2289109313372268, 'epoch': 2.0})

With only one epoch:

what does training loss means:

CPU times: user 5min 11s, sys: 1.4 s, total: 5min 13s
Wall time: 5min 28s
TrainOutput(global_step=515, training_loss=3.020385786630575, metrics={'train_runtime': 327.2815, 'train_samples_per_second': 100.531, 'train_steps_per_second': 1.574, 'total_flos': 906421548064512.0, 'train_loss': 3.020385786630575, 'epoch': 1.0})



Training completed. Do not forget to share your model on huggingface.co/models =)


CPU times: user 10min 40s, sys: 1.47 s, total: 10min 42s
Wall time: 10min 46s
TrainOutput(global_step=1030, training_loss=1.2289109313372268, metrics={'train_runtime': 645.6688, 'train_samples_per_second': 101.916, 'train_steps_per_second': 1.595, 'total_flos': 1807646409192192.0, 'train_loss': 1.2289109313372268, 'epoch': 2.0})

In [ ]:
trainer.push_to_hub("SMILES_RoBERTa")

Saving model checkpoint to SMILES_RoBERTa
Configuration saved in SMILES_RoBERTa/config.json
Model weights saved in SMILES_RoBERTa/pytorch_model.bin
Dropping the following result as it does not have all the necessary fields:
{'task': {'name': 'Masked Language Modeling', 'type': 'fill-mask'}}


model.safetensors:   0%|          | 0.00/199M [00:00<?, ?B/s]

events.out.tfevents.1700668625.55ece987f637.5429.0:   0%|          | 0.00/4.62k [00:00<?, ?B/s]

events.out.tfevents.1700669108.55ece987f637.5429.1:   0%|          | 0.00/4.27k [00:00<?, ?B/s]

Upload 5 LFS files:   0%|          | 0/5 [00:00<?, ?it/s]

events.out.tfevents.1700669659.55ece987f637.5429.2:   0%|          | 0.00/4.78k [00:00<?, ?B/s]

training_args.bin:   0%|          | 0.00/4.54k [00:00<?, ?B/s]

'https://huggingface.co/rebe121314/SMILES_RoBERTa/tree/main/'

In [ ]:
MASK_TOKEN = tokenizer.mask_token


In [ ]:
tokenizer.add_tokens(MASK_TOKEN)


1

In [ ]:
from transformers import pipeline

fill_mask = pipeline(
    "fill-mask",
    model="/content/SMILES_RoBERTa",
    tokenizer = tokenizer
)

MASK_TOKEN = tokenizer.mask_token


loading configuration file /content/SMILES_RoBERTa/config.json
Model config RobertaConfig {
  "_name_or_path": "/content/SMILES_RoBERTa",
  "architectures": [
    "RobertaForMaskedLM"
  ],
  "attention_probs_dropout_prob": 0.1,
  "bos_token_id": 0,
  "classifier_dropout": null,
  "eos_token_id": 2,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "initializer_range": 0.02,
  "intermediate_size": 3072,
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 514,
  "model_type": "roberta",
  "num_attention_heads": 12,
  "num_hidden_layers": 6,
  "pad_token_id": 1,
  "position_embedding_type": "absolute",
  "torch_dtype": "float32",
  "transformers_version": "4.35.2",
  "type_vocab_size": 1,
  "use_cache": true,
  "vocab_size": 8000
}

loading configuration file /content/SMILES_RoBERTa/config.json
Model config RobertaConfig {
  "_name_or_path": "/content/SMILES_RoBERTa",
  "architectures": [
    "RobertaForMaskedLM"
  ],
  "attention_probs_dropout_prob": 0.1,


[{'score': 0.17235350608825684,
  'token': 39,
  'token_str': 'O',
  'sequence': 'La suno O.'},
 {'score': 0.1342533677816391,
  'token': 16,
  'token_str': '1',
  'sequence': 'La suno 1.'},
 {'score': 0.11434822529554367,
  'token': 38,
  'token_str': 'N',
  'sequence': 'La suno N.'},
 {'score': 0.0990724116563797,
  'token': 30,
  'token_str': 'C',
  'sequence': 'La suno C.'},
 {'score': 0.08263091742992401,
  'token': 59,
  'token_str': 'l',
  'sequence': 'La suno l.'}]

In [ ]:

#mask = =O
#CC(C)C1=NC(=CS1)CN(C)C(=O)NC(C(C)C)C(<mask>)NC(CC2=CC=CC=C2)CC(C(CC3=CC=CC=C3)NC(=O)OCC4=CN=CS4)O


[{'score': 0.5472647547721863,
  'token': 17,
  'token_str': '2',
  'sequence': 'CC(C)C1=NC(=CS1)CN(C)C2)CC(C(CC3=CC=CC=C3)NC(=O)OCC4=CN=CS4)O'},
 {'score': 0.131235733628273,
  'token': 18,
  'token_str': '3',
  'sequence': 'CC(C)C1=NC(=CS1)CN(C)C3)CC(C(CC3=CC=CC=C3)NC(=O)OCC4=CN=CS4)O'},
 {'score': 0.09699758887290955,
  'token': 59,
  'token_str': 'l',
  'sequence': 'CC(C)C1=NC(=CS1)CN(C)Cl)CC(C(CC3=CC=CC=C3)NC(=O)OCC4=CN=CS4)O'},
 {'score': 0.05471877381205559,
  'token': 39,
  'token_str': 'O',
  'sequence': 'CC(C)C1=NC(=CS1)CN(C)CO)CC(C(CC3=CC=CC=C3)NC(=O)OCC4=CN=CS4)O'},
 {'score': 0.05185355618596077,
  'token': 16,
  'token_str': '1',
  'sequence': 'CC(C)C1=NC(=CS1)CN(C)C1)CC(C(CC3=CC=CC=C3)NC(=O)OCC4=CN=CS4)O'}]